### PR TITLE
create generic backend to cut dependency on daemon 

### PR DESCRIPTION
**Pull request by MHBauer**
seeing #17463 gave me the idea that at least somewhat the routers can be 'impl-ized' individually.

If we apply similar changes to the `local` router, that would successfully split the api.

This is just an sample of what the final version would look like. I am interested in what people think about splitting the different sections into separate routers and separate `Backends`, vs keeping them all as one, with a single `Backend`.

Names are subject to change, and I need to write a better commit message.

### CODE REVIEW

**Comment by MHBauer** [comment]
@duglin @calavera @icecrime bleh, attention.

**Comment by calavera** [comment]
yep, designing several interfaces was our plan all together :smile_cat: 

/cc @dongluochen

**Comment by calavera** [comment]
PS: the title could be slightly more descriptive.

**Comment by dongluochen** [comment]
This is cool. I'm working on image routes under local.

**Comment by calavera** [comment]
> I'm working on image routes under local.

please, don't touch the image router until #17324 is merged.

**Comment by calavera** [comment]
LGTM

File: api/server/router/network/network.go [hunk 1]
```diff
@@ -6,21 +6,20 @@ import (
...
 	"github.com/docker/docker/api/server/router/local"
-	"github.com/docker/docker/daemon"
 	"github.com/docker/docker/errors"
 	"golang.org/x/net/context"
 )
 
 // networkRouter is a router to talk with the network controller
 type networkRouter struct {
-	daemon *daemon.Daemon
+	impl   Backend
```
Comment (vdemeester) [inline api/server/router/network/network.go]: I feel `impl` is too "generic" name, `backend` should probably be better, or .. I don't know :sweat_smile:

File: api/server/router/network/backend.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package network
+
+import (
+	// network config needs to be pull out to here
+	"github.com/docker/docker/daemon/network"
+
+	"github.com/docker/libnetwork"
+)
+
+// Backend is all the methods that need to be implemented to provide
```
Comment (calavera) [inline api/server/router/network/backend.go]: can we not use the word `stuff` in godocs?

File: api/server/router/network/network.go (hunk 1 above)
Comment (calavera) [inline api/server/router/network/network.go]: can we call this var `backend` rather than `impl`?

File: api/server/router/network/backend.go (hunk 2 above)
Comment (MHBauer) [inline api/server/router/network/backend.go]: Yea, that should probably be a real godoc doc too.
`// Backend represents blah blah blah`
I am kind of surprised none of the checkers caught it. Is this package not linted?

File: api/server/router/network/network.go (hunk 1 above)
Comment (MHBauer) [inline api/server/router/network/network.go]: I want to change the name to whatever people think makes most sense. I had a distinct lack of imagination here. Look for this change in the next version.

File: api/server/router/network/backend.go (hunk 2 above)
Comment (vdemeester) [inline api/server/router/network/backend.go]: Well `Backend` is the name of the struct so it looks fine from the `golint` point of view :sweat_smile:

Comment (MHBauer) [inline api/server/router/network/backend.go]: lol, I missed that line above. Hopefully it looks better now? I'm not sure that comment is terribly content filled, but I gave it another shot.

**Comment by calavera** [comment]
LGTM

**Comment by duglin** [comment]
LGTM

**Pull request merged**